Issue #67 Add support for Minishift integration tests

agajdosi opened the pull request:
Fix #67

budhrg commented:
@jorgemoralespou Any input on such testing framework for minishift addons ?
See https://github.com/minishift/minishift-addons/issues/67

Another PR https://github.com/minishift/minishift-addons/pull/84 which provide platform to addon developer to write shell script based tests. We just run their test in CI.

jorgemoralespou commented:
@agajdosi @budhrg @LalatenduMohanty 
Here's my view.

I like there's a framework for testing addons. This framework should be easy to use and tests easy to write. For both options I struggle with that premise, but the one in this PR could seem more natural, although also more complex to addon developers.

Do I need to run make to test run the tests? Do I need golang or similar on my machine? Works the same on windows, linux or mac hosts? Is it properly documented?
 
I'm looking at the [example](https://github.com/agajdosi/minishift-addons/blob/718a503dbae1a48b9ec1cb21791d4505ea8486d8/test/integration/features/cors.feature) and I'm no real developer, and I would struggle to write this. What happens if there's a typo?

Also, would merging PR be enforced by the presence of tests? Would tests be mandatory?

budhrg commented:
> For both options I struggle with that premise, but the one in this PR could seem more natural, although also more complex to addon developers.

Agree, it is natural and reads well but complex in its underline implementation.
Like 
https://github.com/agajdosi/minishift-addons/blob/718a503dbae1a48b9ec1cb21791d4505ea8486d8/test/integration/features/cors.feature#L12

has underline Go implementation as
https://github.com/minishift/minishift/blob/3f6e5688d69cf8beccf3f45f95900916c1388664/test/integration/minishift.go#L186-L198

Similarly, all the steps in `.feature` file.

> Do I need to run make to test run the tests? 

Yes

> Do I need golang or similar on my machine?

Yes

> Is it properly documented?

Yes, it will be including all the steps available.

> What happens if there's a typo?

Probably test will fail.

> Would tests be mandatory

Yes, it will help to main addons properly. The idea is to run these tests as part of nightly CI job.

praveenkumar commented:
@jorgemoralespou As an add-on developer mostly you need to ensure basic addon functionality like install, uninstall, apply and remove works without any error. That basic feature file will be available for add-on developer and ready to consume. If you want some other logic to test in the addon then only you have to understand golang specific part.

> Also, would merging PR be enforced by the presence of tests? Would tests be mandatory?

Until we have a CI in place for this, once we add job then it will become mandatory.

LalatenduMohanty commented:
> @jorgemoralespou As an add-on developer mostly you need to ensure basic addon functionality like install, uninstall, apply and remove works without any error. That basic feature file will be available for add-on developer and ready to consume. If you want some other logic to test in the addon then only you have to understand golang specific part

+1. @jorgemoralespou  what we are looking for is tests to make sure install, apply remove and uninstall work fine with every release of Minishift and to write these tests you can take a template and modify it. We will/should provide the template.

LalatenduMohanty commented on CONTRIBUTING.adoc:
````diff
@@ -20,6 +20,53 @@ Thank you for your contributions! Please follow this process to submit a patch:
+```
+make integration 
+```
+
````
The prerequisite section is not user-friendly. We need to mention how a user can install dep. I would suggest we putt the steps in bullet points i.e. 
- recent link:[Go] distribution (>=1.8) is installed
- https://docs.openshift.org/latest/minishift/contributing/developing.html#create-go-workspace
- link:https://github.com/golang/dep[dep package] is installed
 -- go get -u github.com/golang/dep/cmd/dep
- Clone this repository. `cd minishift-addons`
- Get latest minishift binary from https://github.com/minishift/minishift/releases/latest and put it in `testing/bin`
-  To get all the go dependency, run  `make vendor`

LalatenduMohanty commented on test/integration/features/cors.feature:
```diff
@@ -0,0 +1,23 @@
+  Scenario: Installing the add-on
+     When file from "https://raw.githubusercontent.com/minishift/minishift-addons/master/add-ons/cors/cors.addon" is downloaded into location "download/cors"
+      And executing "minishift addons install ../../testing/integration-test/download/cors" succeeds
+     Then stdout should contain "Addon 'cors' installed"
```
This does not look intuitive, especially when we are in the addons repository.  We need a way to point to the relative path for the addon. We want to test the PR alos i.e. the add-on would be in the PR itself. So it won't have a physical location in the repository itself.

agajdosi replied:
@LalatenduMohanty +1, will create an issue in the Minishift

agajdosi replied:
@LalatenduMohanty Haha, just realized I can do: `When executing "minishift addons install ../../add-ons/cors" succeeds`. No need for any copying, any download, just dead simple. Thanks for pointing it out, would not realize this.

LalatenduMohanty commented on test/integration/features/cors.feature:
```diff
@@ -0,0 +1,23 @@
+  Scenario: Deleting Minishift
+    Given Minishift has state "Running"
+     When executing "minishift delete --force" succeeds
+     Then Minishift should have state "Does Not Exist"
```
@agajdosi Ideally we should should have an example where we can apply, remove, uninstall the add-on. But maybe it is  another task.

agajdosi replied:
@LalatenduMohanty I would like to have the example in this PR. Ideally it should work I think, so I took the cors add-on, but it can't be removed. Wondering if there is a better candidate for this or if I could implement the `cors.addon.remove`. Then we could just mention it as an example.

Other plan could be to create an example addon, which would do nothing just install, apply, remove and uninstall itself. This file could be used for the add-on documentation also because it would contain all the needed basics which add-ons need. After that I can create an `example.feature` which would show the minimal setup of the feature file to test minimal functionality of add-on.

The second seems fine to me, the question is whether @minishift/minishift-dev are fine with it?

budhrg replied:
> The second seems fine to me, the question is whether @minishift/minishift-dev are fine with it?

+1 for this. `cors` is certainly not the good example. We can add it as part of another PR where we need to add tests.

agajdosi replied:
@budhrg @LalatenduMohanty I have added `example.addon` which contains just `echo` and then `example.feature` which tests this blank addon. It is also used in documentation.

Please note that the example feature has two variant: `example.feature` and `example-clean.feature`, the second one is less commented and I think looks more tidy and readable. I plan we choose and use only one of these variants.

agajdosi commented:
@budhrg @jorgemoralespou 
>>    Do I need to run make to test run the tests?
>
> Yes

One could also use `go test`, but that's not really user friendly:
```
go test -timeout 7200s github.com/minishift-addons/minishift/test/integration --tags=integration -v -args --test-dir /home/agajdosi/golang/src/github.com/minishift/minishift/out/integration-test --binary /home/agajdosi/golang/bin/minishift --tags cors
```

budhrg commented on CONTRIBUTING.adoc:
```diff
@@ -20,6 +20,53 @@ Thank you for your contributions! Please follow this process to submit a patch:
 NOTE: If you submit a new add-on, the add-on must be added as a sub-directory in the *_/add-ons/_* sub-directory of the repository.
 See the documentation for information about link:https://docs.openshift.org/latest/minishift/using/addons.html#add-ons-overview[file structure and content format].
 
+[[testing]]
```
This should be in separate file and should referenced from README. Any additoin to `CONTRIBUTING.MD` will be shown as part of link during new issue creation.

budhrg commented on CONTRIBUTING.adoc:
```diff
@@ -20,6 +20,53 @@ Thank you for your contributions! Please follow this process to submit a patch:
+[[testing]]
+== Testing Add-ons
+
+This repository uses integration test suite from `minishift/minishift` repository to test the functionality of the add-ons.
```
Hyperlink to minishift repo

budhrg commented on CONTRIBUTING.adoc:
````diff
@@ -20,6 +20,53 @@ Thank you for your contributions! Please follow this process to submit a patch:
+    Given Minishift has state "Running"
+     When executing "minishift addons apply cors" succeeds
+     Then stdout should contain "CORS is now allowed from any address"
+```
````
`Note` is needed to following the alignment pattern of `Given`, `When` and `Then` for readability IMO. Otherwise, this would be our first review comment :wink:

budhrg commented on CONTRIBUTING.adoc:
````diff
@@ -20,6 +20,53 @@ Thank you for your contributions! Please follow this process to submit a patch:
+make vendor
+```
+
+=== Running Tests
````
Here, I believe, we should write how to run your addon's test instead of running all features files since the contributor is more interested in his tests.
 
```
=== Running tests for your Addon

Once the [prerequisites](#prerequisites) are fulfilled, run following command to test your addon:
'''
make integration GODOG_OPTS="-tags <addon-name>
'''

Note: `make integration` will run all tests which include tests of other addons.
```

budhrg replied:
To make running tests more intuitive locally, I would prefer following command:
```
make integration ADDON=<addon-name>
```
which will add `GODOG_OPTS="-tags $(ADDON)` under the hood.

if no `ADDON`, then run all tests.

agajdosi commented:
@budhrg Thanks for the feedback, will work on it on Monday.

agajdosi commented on TESTING.adoc:
```diff
@@ -0,0 +1,91 @@
+[[addons-tests]]
```
@robin-owen Please take a look on this file when you have time. I would like to also ask if it feels understandable - as I am quite familiar with Minishift integration tests and Gherkin I feel my opinion here might be affected. Thanks :)

robin-owen commented on TESTING.adoc:
```diff
@@ -0,0 +1,91 @@
+Integration tests are implemented in `test/integration/integration_test.go` file which imports test suite from Minishift repository and can be used for import of additional step definitions.
+
+To help with execution of integration tests `make integration` target is provided in the Makefile.
+Its usage is covered later in this document. 
```
Rather than mention this, could we provide a link to where this is described so that the reader doesn't have to go looking for it manually?

robin-owen commented on TESTING.adoc:
```diff
@@ -0,0 +1,91 @@
+- get latest minishift binary from https://github.com/minishift/minishift/releases/latest and put it into `testing/bin`
+- vendor needed Go packages by runining: `make vendor` in root directory of the `minishift-addons` repository
+
+NOTE: We recommend to vendor the packages once a while to make sure you use up to date packages.
```
How often? "Once in a while" can mean many different spans of time to different people.

robin-owen commented on TESTING.adoc:
```diff
@@ -0,0 +1,91 @@
+[[addons-tests]]
+= Add-ons Tests
```
This is all about integration tests, correct? Could we title this something like "Integration Testing for Add-ons"?

I ask because "Add-ons Tests" is awkward to read. :)

robin-owen commented on TESTING.adoc:
```diff
@@ -0,0 +1,91 @@
+For most add-ons this will mean using one feature file located at `tests/integration/features/<name-of-addon>.feature`.
+
+=== Additional Options
+To start the tests on all addons you can run command:
```
This seems like it belongs in the previous section, "Running Tests".

robin-owen commented on add-ons/example/example.addon:
```diff
@@ -0,0 +1,9 @@
+# Description: Shows basic functionality of an add-on
+
+# Aplication of the add-on
+# several commands can be used: echo, sleep, ssh, docker, oc, openshift
```
We have Minishift documentation which covers [which commands are available](http://artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/using/addons.html#addon-commands) in an add-on. Should we link to that instead of having this list to maintain in case of commands being added or removed in the future?

robin-owen commented on add-ons/example/example.addon:
```diff
@@ -0,0 +1,9 @@
+# Name: example
+# Description: Shows basic functionality of an add-on
+
+# Aplication of the add-on
```
There's a typo in "Application." :)

robin-owen commented on CONTRIBUTING.adoc:
```diff
@@ -20,6 +20,8 @@ Thank you for your contributions! Please follow this process to submit a patch:
 NOTE: If you submit a new add-on, the add-on must be added as a sub-directory in the *_/add-ons/_* sub-directory of the repository.
 See the documentation for information about link:https://docs.openshift.org/latest/minishift/using/addons.html#add-ons-overview[file structure and content format].
 
+NOTE: We also recommend to provide an integration test for the add-on. To get more information see link:TESTING.adoc[integration tests documentation]. 
```
```adoc
NOTE: We also recommend providing integration tests for the add-on.
See the link:TESTING.adoc[integration tests documentation] for more information.
```

robin-owen commented:
@agajdosi Yes, that's what I was thinking regarding the workflow. Thanks for clarifying! :)

agajdosi commented:
> We have Minishift documentation which covers which commands are available in an add-on. Should we link to that instead of having this list to maintain in case of commands being added or removed in the future?

@minishift/minishift-dev This PR is ready for review (contains just minor grammatical problems which @robin-owen will fix soon). Are we fine with having the `example.addon` and `example.feature`? Does the documentation seem fine to you?

@robin-owen Changes implemented. Please check and feel free to create the PR for the grammar fixes. Thanks :)

agajdosi commented:
@minishift/minishift-dev Please take a look, whether this PR needs any changes. Thanks :)

budhrg commented on TESTING.adoc:
```diff
@@ -0,0 +1,107 @@
+- Add-on uninstallation: `executing "minishift addons uninstall <name-of-addon>" succeeds`
+- Deletion of the Minishift instance: `executing "minishift delete --force" succeeds`
+
+[[advanced-steps]]
```
This is what I was talking about documentation. 
I would be happy if we list all possible steps one can use to encourage the add-on developers to write tests.
I have a hard time in assuming users will go to testsuite package and read the go code(also regex magic) to see all the available steps.

It should have been in a new doc file.

agajdosi replied:
@budhrg Agree. I am thinking about it for some time even at Minishift repository, I would like to take it as part of https://github.com/minishift/minishift/issues/1423. However for this PR I think we could be fine without it - still one can take a look on existing feature files and re-use parts from them. I can add more feature files for addons we currently have in repo and try to test more deeply where I would be able to do so. Maybe this could help the contributors, wdyt?

budhrg replied:
> I am thinking about it for some time even at Minishift repository, I would like to take it as part of minishift/minishift#1423

That is different issue altogether I guess. Issue 1423 is to document tags. 
We can create another issue to document all available steps.

> I can add more feature files for addons we currently have in repo.

This is not part of this issue I guess and for some addon it might take some more time like Istio.

budhrg commented on test/integration/integration_test.go:
```diff
@@ -0,0 +1,64 @@
+// +build integration
+
+/*
+Copyright (C) 2017 Red Hat, Inc.
```
I think we can remove this license. In Minishift we had some check which made us to use it.

budhrg commented on TESTING.adoc:
```diff
@@ -0,0 +1,107 @@
+Create a new file named *_tests/integration/features/<name-of-add-on>.feature_* to create a new integration test for an add-on.
+The *_example.feature_* file demonstrates the basic structure of an integration test for an example add-on and may be used as a starting point.
+
+Features should contain at least 6 key parts which secure that the tested add-on can be installed, applied, removed and uninstalled:
```
@jorgemoralespou Could you look into this documentation on writing tests. Also see `Advanced steps`. There is a link to https://github.com/minishift/minishift/tree/master/test/integration/testsuite for more reference.

agajdosi commented:
Hi @minishift/minishift-dev, could you please take a look on this PR?

agajdosi commented:
@LalatenduMohanty Great, thanks. Let me know if you have any questions on the usage.

LalatenduMohanty commented:
@agajdosi The PR does not work for me

```
$ make integration MINISHIFT_BINARY=/Users/lmohanty/code/gocode/bin/minishift
mkdir -p /Users/lmohanty/code/gocode/src/github.com/minishift/minishift-addons/testing/integration-test
go test -timeout 3600s github.com/minishift/minishift-addons/test/integration --tags=integration -v -args --test-dir /Users/lmohanty/code/gocode/src/github.com/minishift/minishift-addons/testing/integration-test --binary /Users/lmohanty/code/gocode/bin/minishift \
	--run-before-feature="" --test-with-specified-shell="" --tags=""
Test run using Boot2Docker iso image.
Log successfully started, logging into: /Users/lmohanty/code/gocode/src/github.com/minishift/minishift-addons/testing/test-results/integration_2018-3-9_17-20-00.log
Running Integration test in: /Users/lmohanty/code/gocode/src/github.com/minishift/minishift-addons/testing/integration-test
Using binary: /Users/lmohanty/code/gocode/bin/minishift
Feature: Cors add-on
  As Minishift user I can install and apply Cors add-on from minishift-addons repository.

  Scenario: Installing the add-on                                         # features/cors.feature:5
    When executing "minishift addons install ../../add-ons/cors" succeeds # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "Addon 'cors' installed"                   # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain

  Scenario: Minishift starts                   # features/cors.feature:9
    Given Minishift has state "Does Not Exist" # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState
    When executing "minishift start" succeeds  # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then Minishift should have state "Running" # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState

  Scenario: Applying the add-on                                       # features/cors.feature:14
    When executing "minishift addons apply cors" succeeds             # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "CORS is now allowed from any address" # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain

  Scenario: Deleting Minishift                         # features/cors.feature:18
    Given Minishift has state "Running"                # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState
    When executing "minishift delete --force" succeeds # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Command 'delete --force' did not execute successfully. cmdExit: 1, cmdErr: Error deleting '/Users/lmohanty/code/gocode/src/github.com/minishift/minishift-addons/testing/integration-test/machines': remove /Users/lmohanty/code/gocode/src/github.com/minishift/minishift-addons/testing/integration-test/machines/minishift/root-volume.sparsebundle/bands/135: permission denied

    Then Minishift should have state "Does Not Exist"  # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState

Feature: Example feature for example.addon

  Scenario: User can install example add-on                                    # features/example-clean.feature:7
    Given stdout of command "minishift addons list" does not contain "example" # testsuite.go:520 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnContains
    When executing "minishift addons install ../../add-ons/example" succeeds   # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Command 'addons install ../../add-ons/example' did not execute successfully. cmdExit: 1, cmdErr: Add-on installation failed with the error: Unable to parse specified addon: Add-on directory name should match to addon name
xxxxxxxxxxxxxxxxxx
```

agajdosi commented:
Hi @LalatenduMohanty, thanks for noticing, there was a type in `example.addon.uninstall` - name of addon was with capital letter. Fixed now and works for me now.

LalatenduMohanty commented:
@agajdosi I am still seeing some issues. Will take a fresh look tomorrow.

agajdosi commented:
Hi @LalatenduMohanty, I have updated the Gopkg.lock file, added the `make clean` step and it passes fine both example and core.feature for me on a Mac machine with minishift-v1.14.0 binary, here's the core run:

```gherkin
cdkqe-mac:minishift-addons cdkqe$ make integration ADDON=cors
mkdir -p /Users/cdkqe/go/src/github.com/minishift/minishift-addons/testing/integration-test
go test -timeout 3600s github.com/minishift/minishift-addons/test/integration --tags=integration -v -args --test-dir /Users/cdkqe/go/src/github.com/minishift/minishift-addons/testing/integration-test --binary /Users/cdkqe/go/src/github.com/minishift/minishift-addons/testing/bin/minishift \
	--run-before-feature="" --test-with-specified-shell="" --tags=cors 
Test run using Boot2Docker iso image.
Log successfully started, logging into: /Users/cdkqe/go/src/github.com/minishift/minishift-addons/testing/test-results/integration_2018-3-16_15-42-34.log
Running Integration test in: /Users/cdkqe/go/src/github.com/minishift/minishift-addons/testing/integration-test
Using binary: /Users/cdkqe/go/src/github.com/minishift/minishift-addons/testing/bin/minishift
Feature: Cors add-on
  As Minishift user I can install and apply Cors add-on from minishift-addons repository.

  Scenario: Installing the add-on                                           # features/cors.feature:5
    Given stdout of command "minishift addons list" does not contain "cors" # testsuite.go:520 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnContains
    When executing "minishift addons install ../../add-ons/cors" succeeds   # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "Addon 'cors' installed"                     # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain

  Scenario: Minishift starts                   # features/cors.feature:10
    Given Minishift has state "Does Not Exist" # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState
    When executing "minishift start" succeeds  # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then Minishift should have state "Running" # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState

  Scenario: Applying the add-on                                       # features/cors.feature:15
    When executing "minishift addons apply cors" succeeds             # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "CORS is now allowed from any address" # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain

  Scenario: Deleting Minishift                         # features/cors.feature:19
    Given Minishift has state "Running"                # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState
    When executing "minishift delete --force" succeeds # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then Minishift should have state "Does Not Exist"  # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState

Feature: Example feature for example.addon
Error closing the log: invalid argument

4 scenarios (4 passed)
11 steps (11 passed)
5m43.619265033s
testing: warning: no tests to run
PASS
ok  	github.com/minishift/minishift-addons/test/integration	343.665s [no tests to run]
cdkqe-mac:minishift-addons cdkqe$ ```

agajdosi commented:
@LalatenduMohanty I have used `go 1.10` on Mac, but that should not have much impact. I am still thinking about the `GITHUB_API_TOKEN` you had on the machine, from what I have found in docs and what I use everywhere for the minishift it should be: `MINISHIFT_GITHUB_API_TOKEN`, for example: https://github.com/minishift/minishift/issues/1920. The startup can fail for you due to the API limit, while normal startups are fine because the oc is already cached and is not downloaded. Unfortunately I do not have any better idea why it fails to you :sob:

budhrg commented on Makefile:
```diff
@@ -0,0 +1,48 @@
+
+# Make target definitions
+.PHONY: integration
+integration: ADDON ?= ""
```
Need `vendor` as dependent target for `integration`.

agajdosi replied:
+1

budhrg commented on TESTING.adoc:
````diff
@@ -0,0 +1,107 @@
+For example:
+
+```
+make integration ADDON=xpaas MINISHIFT_BINARY=/home/user/minishift/minishift
````
It is high chance that user will clone this repo in `$GOPATH/src/minishift/` path.
Let's add the following as part of the documentation
```
$ git clone https://github.com/minishift/minishift-addons $GOPATH/src/github.com/minishift/minishift-addons
```
as I was getting following errors on initial run(since I had clone it in different location):
```
$ make integration MINISHIFT_BINARY=/Users/devtools/gowork/bin/minishift
mkdir -p /Users/devtools/redhat/minishift-addons/testing/integration-test
go test -timeout 3600s github.com/minishift/minishift-addons/test/integration --tags=integration -v -args --test-dir /Users/devtools/redhat/minishift-addons/testing/integration-test --binary /Users/devtools/gowork/bin/minishift \
	--run-before-feature="" --test-with-specified-shell="" --tags=""
can't load package: package github.com/minishift/minishift-addons/test/integration: cannot find package "github.com/minishift/minishift-addons/test/integration" in any of:
	/Users/devtools/go/src/github.com/minishift/minishift-addons/test/integration (from $GOROOT)
	/Users/devtools/gowork/src/github.com/minishift/minishift-addons/test/integration (from $GOPATH)
make: *** [integration] Error 1
```

agajdosi replied:
Thanks @budhrg for noticing. Going now to implement the suggestions.

agajdosi replied:
@budhrg Actually just realized, this topic is covered in "Prerequisites" chapter:
```
- `git clone` this repository into *_$GOPATH/src/github.com/minishift/minishift-addons_*.
```

agajdosi commented:
@budhrg @LalatenduMohanty suggestions implemented.

budhrg commented on Makefile:
```diff
@@ -0,0 +1,51 @@
+MINISHIFT_BINARY ?= $(TEST_DIR)/bin/minishift$(IS_EXE)
+
+# Make target definitions
+target: ;
```
@agajdosi , use below as `make` output is not intuitive :smile: 

```
default:
               @echo "Nothing to make. Run 'make integration' to run integration tests."
```

LalatenduMohanty commented on Makefile:
```diff
@@ -0,0 +1,52 @@
+	go test -timeout $(TIMEOUT) $(REPOPATH)/test/integration --tags=integration -v -args --test-dir $(TEST_DIR)/integration-test --binary $(MINISHIFT_BINARY) \
+	--run-before-feature="$(RUN_BEFORE_FEATURE)" --test-with-specified-shell="$(TEST_WITH_SPECIFIED_SHELL)" --tags=$(ADDON) $(GODOG_OPTS)
+
+.PHONY: vendor
```
@agajdosi @budhrg We need to remove this line i.e. ` .PHONY: vendor` as with this `make integration` will always run `dep ensure -v` but if we already have the directory we do not need to run that.

LalatenduMohanty commented:
Tests are passing on my macbook

```
eature: Cors add-on
  As Minishift user I can install and apply Cors add-on from minishift-addons repository.

  Scenario: Installing the add-on                                           # features/cors.feature:5
    Given stdout of command "minishift addons list" does not contain "cors" # testsuite.go:520 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnContains
    When executing "minishift addons install ../../add-ons/cors" succeeds   # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "Addon 'cors' installed"                     # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain

  Scenario: Minishift starts                   # features/cors.feature:10
    Given Minishift has state "Does Not Exist" # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState
    When executing "minishift start" succeeds  # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then Minishift should have state "Running" # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState

  Scenario: Applying the add-on                                       # features/cors.feature:15
    When executing "minishift addons apply cors" succeeds             # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "CORS is now allowed from any address" # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain

  Scenario: Deleting Minishift                         # features/cors.feature:19
    Given Minishift has state "Running"                # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState
    When executing "minishift delete --force" succeeds # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then Minishift should have state "Does Not Exist"  # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState

Feature: Example feature for example.addon

  Scenario: User can install example add-on                                    # features/example.feature:7
    Given stdout of command "minishift addons list" does not contain "example" # testsuite.go:520 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnContains
    When executing "minishift addons install ../../add-ons/example" succeeds   # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout of command "minishift addons list" contains "example"          # testsuite.go:520 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnContains

  Scenario: User can apply the example add-on                        # features/example.feature:12
    Given executing "minishift start" succeeds                       # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    When executing "minishift addons apply example" succeeds         # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "Example add-on successfully applied" # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain

  Scenario: User can remove the example add-on                       # features/example.feature:17
    When executing "minishift addons remove example" succeeds        # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "Example add-on successfully removed" # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain

  Scenario: User can uninstall example add-on                                # features/example.feature:21
    When executing "minishift addons uninstall example" succeeds             # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "Add-on 'example' uninstalled"                # testsuite.go:529 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnShouldContain
    And stdout of command "minishift addons list" does not contain "example" # testsuite.go:520 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.commandReturnContains

  Scenario: Deleting Minishift                         # features/example.feature:26
    When executing "minishift delete --force" succeeds # testsuite.go:587 -> github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.executingMinishiftCommandSucceedsOrFails
    Then Minishift should have state "Does Not Exist"  # testsuite.go:76 -> *Minishift.github.com/minishift/minishift-addons/vendor/github.com/minishift/minishift/test/integration/testsuite.shouldHaveState
Error closing the log: invalid argument

9 scenarios (9 passed)
24 steps (24 passed)
5m48.960866s
testing: warning: no tests to run
PASS
ok  	github.com/minishift/minishift-addons/test/integration	349.089s
```

LalatenduMohanty commented:
Merged 104bd998ba4f35c84910ffe34100b4b57f0ae0cd

LalatenduMohanty commented:
Thanks @agajdosi :-)
